File: 0260_Single_Number_III/solution.java
package cn.huang.leetcode;

import java.util.HashMap;

/*

260. Single Number III

Given an array of numbers nums, in which exactly two elements appear only once and all the other elements appear
exactly twice. Find the two elements that appear only once.

Example:

Input:  [1,2,1,3,2,5]
Output: [3,5]

Note:

    The order of the result is not important. So in the above example, [5, 3] is also correct.
    Your algorithm should run in linear runtime complexity. Could you implement it using only constant space complexity?

 */
public class LC_0260_SingleNumber_III {
    public int[] singleNumber(int[] nums) {
        HashMap<Integer, Integer> numbers = new HashMap<>();
        for(int i = 0; i < nums.length; i++) {
            if( numbers.containsKey(nums[i]) == false ) {
                numbers.put(nums[i], 1);
            } else {
                numbers.remove(nums[i]);
            }
        }

        int[] target = new int[numbers.size()];
        int i = 0;
        for (Integer key : numbers.keySet()) {
            target[i] = key.intValue();
            i++;
        }

        return target;
    }
}

File: 0260_Single_Number_III/solution.cpp
/*
260. Single Number III

Level: Medium

https://leetcode.com/problems/single-number-iii
*/
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <numeric>
#include <algorithm>

#include "common/ListNode.h"
#include "common/TreeNode.h"

using namespace std;

/*
 two elements appear only once and all the other elements appear exactly twice.
*/
/*
  Solution: 
  https://www.youtube.com/watch?v=3TSC0nlur58
  
*/

class Solution
{
public:
  vector<int> singleNumber(vector<int> &nums)
  {
    int axorb = 0;
    for (int n : nums)
      axorb ^= n; // 得到的是只出现了一次的两个数字的异或。
    
    int a = 0, b = 0;
    int mask = 1;
    // find out the lowest 1 in the axorb
    while ((mask & axorb) == 0)
      mask <<= 1;

    for (int n : nums)
    {
      if (n & mask)
        a ^= n;
      else
        b ^= n;
    }
    return {a, b};
  }
};